### PR TITLE
Add missing type promotions for pythonapi

### PR DESCRIPTION
**Pull request by MarkVeerasingam**
This PR resolves issue #9640 by adding a type promotion checker that would check the integer bit width and apply a signext or zext attribute in `numba/core/cpu.py`. Aswell, inside `numba/core/pythonapi.py` for C lowering and `numba/core/base.py`, a no-op hook to call this in declare_function.

The IBM Z ABI requires all integer types to be promoted to full register width by the caller. The callee will rely on this promotion to be done by the caller and will use an int32 input directly as 64 bit value in instructions without applying additional sign- or zeroextensions to it.

Without type promotion checking, Numba on s390x would segfault, as observed in https://github.com/numba/llvmlite/issues/780. 

CC: @Andreas-Krebbel

### CODE REVIEW

**Comment by gmarkall** [comment]
Thanks for the PR! Do you have a link to a document specifying the s390x ABI please?

**Comment by MarkVeerasingam** [comment]
> Thanks for the PR! Do you have a link to a document specifying the s390x ABI please?

Thanks!
This repo has the general s390x ABI: https://github.com/IBM/s390x-abi/

You can find the latest releases of the ABI here: https://github.com/IBM/s390x-abi/releases/

**Comment by gmarkall** [comment]
The build logs have disappeared now, but I think the fail was flake8-related - could you investigate the issue and resolve it once the CI run has completed again please?

This will also need a release note.

**Comment by MarkVeerasingam** [comment]
> The build logs have disappeared now, but I think the fail was flake8-related - could you investigate the issue and resolve it once the CI run has completed again please?
> 
> This will also need a release note.

Thanks! I created a release note called `10396.bug_fix.rst` in `numba/docs/upcoming_changes`.
I just resolved the Flake8 linting errors in `numba/core/cpu.py` I've tested the fix locally:

```
(py313) [root@a3e16064 numba]# flake8 numba/core/cpu.py
(py313) [root@a3e16064 numba]#
```

latest push should now pass all CI checks.

File: numba/core/base.py [hunk 1]
```diff
@@ -412,6 +412,8 @@ def get_external_function_type(self, fndesc):
     def declare_function(self, module, fndesc):
         fnty = self.call_conv.get_function_type(fndesc.restype, fndesc.argtypes)
         fn = cgutils.get_or_insert_function(module, fnty, fndesc.mangled_name)
+        # Call the target-specific attribute hook. Defined as a no-op, overwrriden in numba/core/cpu.py
```
Comment (gmarkall) [inline numba/core/base.py]: I think this comment is probably unnecessary, it seems to just state what the next line does, and references a far-away implementation detail that could change independently:

```suggestion
```

File: numba/core/cpu.py [hunk 2]
```diff
@@ -50,13 +50,50 @@ def init(self):
         self.is32bit = (utils.MACHINE_BITS == 32)
         self._internal_codegen = codegen.JITCPUCodegen("numba.exec")
 
-        # Add ARM ABI functions from libgcc_s
-        if platform.machine() == 'armv7l':
+        # Add ARM or s390x ABI functions from libgcc_s
+        if platform.machine() == 'armv7l' or platform.machine() == 's390x':
```
Comment (gmarkall) [inline numba/core/cpu.py]: I think armv7l is unsupported now, so this could be:

```suggestion
        # Add s390x ABI functions from libgcc_s
        if platform.machine() == 's390x':
```

**Comment by MarkVeerasingam** [comment]
Thanks for the review!

Regarding your questions
>     * Does this fully supersede #10320?

Yes, this PR supersedes PR https://github.com/numba/numba/pull/10320 in regards to volume of code and work needed for s390x support. However doesnt supersede all issues we encountered for supporting numba for s390x.

We identified 3 areas that needed support, the former PR, covers over all of them but not in a maintainable way:

1. Type Promotions (This PR)
2. 16-Byte SIMD Array Alignment (Opened an issue here https://github.com/numba/numba/issues/10407)
3. Endianness problem in Siphash hashing in ```numba/cpython/hashing.py``` mostly posx, posy errors, I will be opening a dedicated separate PR soon.

>     * The target attributes are applied in two places in this PR. However, there are many uses of `cgutils.get_or_insert_function()` all over the Numba codebase (e.g. parfors, list and dict objects, and a bunch of others) - are the applications of target-specific attributes sufficient?

I just confirmed this, and you are correct, the hook in declare_functions was not necessary. Local tests on s390x still pass with the logic localized to pythonapi.py.

As noted in #9640, pythonapi.py is the critical for these declarations during IR generation and lowering. In the previous PR (#10320), types were manually reassigned from 32-bit to 64-bit in multiple locations. With the promotion logic now handled in this PR, those manual overrides are no longer needed, suggesting that this implementation correctly captures the ABI requirements where they matter most.

**Comment by gmarkall** [comment]
Thanks for clarifying and updating the PR. From the changes and your comments, I think I now understand that applying the attributes matters when calling external functions only, and Numba being internally consistent with respect to calls internal to jitted code is OK even if it's not following the standard ABI. (correct me if I am wrong here)

With this in mind, I'd expect there to be issues calling things through ctypes, cffi, and Cython functions via their address - would you also expect that, or is there something I'm missing here? If there will be issues there, I think this PR is still OK in its present form as a fix for one particular manifestation of the issue, and other items could be addressed in follow-ups.

**Comment by gmarkall** [comment]
@MarkVeerasingam Could you please avoid force-pushes now we're into the review process? Force-pushes make it really hard to track what's going on with the comments and changes through the review (I think because Github doesn't track everything as well once a force-push has happened)

**Comment by MarkVeerasingam** [comment]
> @MarkVeerasingam Could you please avoid force-pushes now we're into the review process? Force-pushes make it really hard to track what's going on with the comments and changes through the review (I think because Github doesn't track everything as well once a force-push has happened)

My apologies for the force-push, I will stick to regular commits from from her on! I was trying to keep the history tidy after realizing the BaseContext cleanup caused regressions in the DPU tests, but I understand it makes the review harder to track. 

I have restored the apply_target_attributes hook in base.py to keep the DPUContext and other extensions happy, so the CI should return to the expected state.

 apologies for any inconvience, will be aware going forward! :)

**Comment by gmarkall** [comment]
Thanks for this, I hadn't even realised that the DPU was affected by removing that hook (I think I missed it due to the force push :slightly_smiling_face:), so thanks for clarifying.

**Review by gmarkall — APPROVED** [review]

**Comment by gmarkall** [comment]
> With this in mind, I'd expect there to be issues calling things through ctypes, cffi, and Cython functions via their address - would you also expect that, or is there something I'm missing here? If there will be issues there, I think this PR is still OK in its present form as a fix for one particular manifestation of the issue, and other items could be addressed in follow-ups.

@MarkVeerasingam Not sure if this part of the discussion got lost in amongst the other comments - do you think a follow-up would be needed for these kinds of calls as well?

**Comment by MarkVeerasingam** [comment]
> > With this in mind, I'd expect there to be issues calling things through ctypes, cffi, and Cython functions via their address - would you also expect that, or is there something I'm missing here? If there will be issues there, I think this PR is still OK in its present form as a fix for one particular manifestation of the issue, and other items could be addressed in follow-ups.

Thanks for this question, I just ssh'd into my Mainframe terminal, to check this, as I have assumed similar during development, we share the same thoughts here.

Context:
I am using all the commit changes in this PR, only adding a 16 byte alignment in base.py for s390x vector facility simd (128 bits) 
```
(py313) [root@a3e16064 numba]# git diff
diff --git a/numba/core/base.py b/numba/core/base.py
index 4c7aa7871..854c1948e 100644
--- a/numba/core/base.py
+++ b/numba/core/base.py
@@ -1130,7 +1130,7 @@ class BaseContext(object):
         Get preferred array alignment for Numba type *ty*.
         """
         # AVX prefers 32-byte alignment
-        return 32
+        return 16

     def post_lowering(self, mod, library):
         """Run target specific post-lowering transformation here.
(py313) [root@a3e16064 numba]#
```

ctypes:
- 14 passed, 1 skipped in 2.23s
- numba/tests/test_ctypes.py::TestCTypesUseCases::test_stdcall SKIPPED (Windows-specific test) 

cffi
just installed cffi from pip Version: 2.0.0
```
(py313) [root@a3e16064 numba]# pytest -v numba/tests/test_cffi.py
============================================================================================== test session starts ==============================================================================================
platform linux -- Python 3.13.2, pytest-9.0.2, pluggy-1.6.0 -- /root/miniconda3/envs/py313/bin/python3.13
cachedir: .pytest_cache
rootdir: /root/numba_4/numba
plugins: xdist-3.8.0, anyio-4.12.1
collected 17 items

numba/tests/test_cffi.py::TestCFFI::test_bool_function_ool PASSED                                                                                                                                         [  5%]
numba/tests/test_cffi.py::TestCFFI::test_from_buffer_error PASSED                                                                                                                                         [ 11%]
numba/tests/test_cffi.py::TestCFFI::test_from_buffer_float32 Segmentation fault (core dumped)
```

From just regular tests ctypes looks okay and from my time working on it, I havent seen any problems or major edge cases that stand out. I am unsure how cffi is used in Numba but by the looks of it, yes, a follow-up would be needed for how cffi is called and promoted.

I couldnt find any immediate cpython tests, if there are any you could point me towards or similar that would be great!

**Comment by gmarkall** [comment]
> I couldnt find any immediate cpython tests, if there are any you could point me towards or similar that would be great!

The calls into the CPython API are being exercised in many places, so if there were an issue I think it'd show up as large-scale failures in many parts of the test suite. I'm not aware of anything specifically testing those calls, but as long as you've run the whole test suite locally I'd have faith in that part of the changes.

**Comment by gmarkall** [comment]
> From just regular tests ctypes looks okay and from my time working on it, I havent seen any problems or major edge cases that stand out.

I wonder if you implement a C function that has a prototype like

```C
void print_numbers(uint8_t a, uint16_t b, uint32_t c, uint64_t d);
```

that prints out each of its parameters, and called it inside an `@njit` function via ctypes, if it would end up printing junk from the upper bits of the parameters, at least some of the time?

**Comment by MarkVeerasingam** [comment]
> > I couldnt find any immediate cpython tests, if there are any you could point me towards or similar that would be great!
> 
> The calls into the CPython API are being exercised in many places, so if there were an issue I think it'd show up as large-scale failures in many parts of the test suite. I'm not aware of anything specifically testing those calls, but as long as you've run the whole test suite locally I'd have faith in that part of the changes.

Thank you! from what we've seen, the test suite is actually quite clean on passes! The former s390x PR https://github.com/numba/numba/pull/10320/files covered everything that needed to be promoted to satisfy the numba suite.
Using this PR the only one that notably stands out is ```numba/cpython/hashing.py``` 

This again is a separate problem in of itself that I want to look into, whether it is a cpython specific calling promotion or just how siphash works. The change below from the former PR read the key in as bytes and reversed them to satisfy BE.

However from my time on this PR changes, I havent seen any other cpython problems outside of the hashing (which i believe is just a siphash specific thing :) 

https://github.com/numba/numba/blob/35a49f783085677a1c7160df60927b650ccd5395/numba/cpython/hashing.py#L445-L470

**Comment by gmarkall** [comment]
@MarkVeerasingam I'm inclined to merge this and await follow-up PRs for the other items. Does that work for you?

**Comment by MarkVeerasingam** [comment]
> @MarkVeerasingam I'm inclined to merge this and await follow-up PRs for the other items. Does that work for you?

That works with me! :)
